File: Day-6/idempotency-sample/python/idempotency_store.py
from __future__ import annotations

from dataclasses import dataclass
from threading import Lock
from typing import Callable, Dict, Generic, TypeVar

T = TypeVar("T")


@dataclass(frozen=True)
class ProcessResult(Generic[T]):
    value: T
    replayed: bool


class IdempotencyStore(Generic[T]):
    def __init__(self) -> None:
        self._responses: Dict[str, T] = {}
        self._lock = Lock()

    def execute(self, key: str, handler: Callable[[], T]) -> ProcessResult[T]:
        if not key:
            raise ValueError("key must be non-empty")

        with self._lock:
            if key in self._responses:
                return ProcessResult(value=self._responses[key], replayed=True)

            value = handler()
            self._responses[key] = value
            return ProcessResult(value=value, replayed=False)

    def processed_keys(self) -> int:
        with self._lock:
            return len(self._responses)

File: Day-6/idempotency-sample/java/IdempotencyStore.java
import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

public class IdempotencyStore<T> {
    private final Map<String, T> responses = new HashMap<>();

    public synchronized ProcessResult<T> execute(String key, Supplier<T> handler) {
        if (key == null || key.isBlank()) {
            throw new IllegalArgumentException("key must be non-empty");
        }

        if (responses.containsKey(key)) {
            return new ProcessResult<>(responses.get(key), true);
        }

        T value = handler.get();
        responses.put(key, value);
        return new ProcessResult<>(value, false);
    }

    public synchronized int processedKeys() {
        return responses.size();
    }

    public record ProcessResult<T>(T value, boolean replayed) {
    }
}
